### PR TITLE
UHK-agent: Support for Ultimate Hacking Keyboard udev-rules and configuration application

### PR DESCRIPTION
**Pull request by ngiger**
###### Motivation for this change

I am using an Ultimate Hacking Keyboard and got today a new trackball module for it, which needed an updated firmware version of the driver (Old one stems probably from my debian system before I switched to Nixos).

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Relase notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

I am unsure whether and where I have to add more documentation about it, eg. to document the needed settings in the configuratiuon.nix:
* services.udev.packages = with pkgs; [ uhk-agent ];
* users.extraGroups.input.members = [ "myUserName" ];

### CODE REVIEW

**Review by bbjubjub2494 — CHANGES_REQUESTED** [review]
Welcome aboard

Result of `nixpkgs-review pr 132420` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>uhk-agent</li>
  </ul>
</details>

> I am unsure whether and where I have to add more documentation about it, eg. to document the needed settings in the configuratiuon.nix

I think the best approach here is code rather than documentation: a simple NixOS module with a high-level option, similar to [what I did for one of my keeb](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/hardware/keyboard/teck.nix). This can also be pushed back to a future PR.

File: pkgs/os-specific/linux/uhk-agent/default.nix [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    install -m 644 -D ${appimageContents}/resources/rules/50-uhk60.rules $out/rules/50-uhk60.rules
+    substituteInPlace $out/share/applications/${pname}.desktop \
+      --replace 'Exec=AppRun' 'Exec=${pname}'
+    cp -r ${appimageContents}/usr/share/icons $out/share
+  '';
+
+  meta = with lib; {
+    description = "Agent is the configuration application of the Ultimate Hacking Keyboard.";
+    homepage = "https://github.com/UltimateHackingKeyboard/agent";
+    license = licenses.mit;
```
Comment (lourkeur) [inline pkgs/os-specific/linux/uhk-agent/default.nix:36]:
```suggestion
    license = licenses.unfreeRedistributable;
```

I should have been more clear: the benneti derivation you credit is MIT, [upstream](https://github.com/UltimateHackingKeyboard/agent/blob/master/LICENSE) isn't MIT.

File: pkgs/os-specific/linux/uhk-udev-rules/default.nix [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+  meta = {
+    description = ''
+    udev rules for UHK keyboards
+    from https://ultimatehackingkeyboard.com.
+    Users must be in the input group. E.g. \n
+    users.extraGroups.input.members = [ "userName" ];\n
+    '';
+    inherit (uhk-agent.meta) license;
+    maintainers = [ lib.maintainers.lourkeur ];
```
Comment (lourkeur) [inline pkgs/os-specific/linux/uhk-udev-rules/default.nix:21]:
```suggestion
    maintainers = [ lib.maintainers.ngiger ];
```

**Review by bbjubjub2494 — CHANGES_REQUESTED** [review]

File: pkgs/os-specific/linux/uhk-udev-rules/default.nix [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+  installPhase = ''
+     install -D -m 644 ${uhk-agent.out}/rules/50-uhk60.rules $out/lib/udev/rules.d/50-uhk60.rules
+  '';
+
+  meta = {
+    description = ''
+    udev rules for UHK keyboards
+    from https://ultimatehackingkeyboard.com.
+    Users must be in the input group. E.g. \n
+    users.extraGroups.input.members = [ "userName" ];\n
```
Comment (lourkeur) [inline pkgs/os-specific/linux/uhk-udev-rules/default.nix:18]:
```suggestion
```
I don't think that's the right place to put it. It also doesn't need to explain how to configure groups. (besides, extraGroups has been renamed to groups and some administrators such as myself prefer `users.users.<name>.extraGroups`)

File: nixos/modules/hardware/keyboard/uhk.nix [hunk 4]
```diff
@@ -0,0 +1,16 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+let
+  cfg = config.hardware.keyboard.uhk;
+in
+{
+  options.hardware.keyboard.uhk = {
+    enable = mkEnableOption "non-root access to the firmware of UHK keyboards";
```
Comment (lourkeur) [inline nixos/modules/hardware/keyboard/uhk.nix:9]:
```suggestion
    enable = mkEnableOption ''
    non-root access to the firmware of UHK keyboards.
    You need it when you want to flash a new firmware on the keyboard.
    Access to the keyboard is granted to users in the "input" group.
    You may want to install the uhk-agent package.
    '';
```
Here seems like a better spot to put it (adapted from `hardware.zsa`)

**Review by bbjubjub2494 — CHANGES_REQUESTED** [review]
sorry for the double review

File: pkgs/os-specific/linux/uhk-udev-rules/default.nix (hunk 2 above)
Comment (ngiger) [reply pkgs/os-specific/linux/uhk-udev-rules/default.nix]: Thanks for spotting this error. I would like to apologize for not thinking enough about copy/pasting code!

Comment (lourkeur) [reply pkgs/os-specific/linux/uhk-udev-rules/default.nix]: Apology accepted! I wasn't sure how to react at first but I figured it was probably a copy/paste. It's no big deal.

File: pkgs/os-specific/linux/uhk-udev-rules/default.nix [hunk 5]
```diff
@@ -0,0 +1,21 @@
+{ lib, stdenv, uhk-agent, fetchurl }:
+
+stdenv.mkDerivation {
+  pname = "uhk-udev-rules";
+  version = lib.getVersion uhk-agent;
+  inherit (uhk-agent);
```
Comment (lourkeur) [inline pkgs/os-specific/linux/uhk-udev-rules/default.nix:6]:
```suggestion
```
redundant

**Review by bbjubjub2494 — CHANGES_REQUESTED** [review]
last bits of lint, thank you for your patience.

Result of `nixpkgs-review pr 132420` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package blacklisted:</summary>
  <ul>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>3 packages built:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>uhk-agent</li>
    <li>uhk-udev-rules</li>
  </ul>
</details>

**Review by bbjubjub2494 — APPROVED** [review]
LGTM

Result of `nixpkgs-review pr 132420` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages built:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>uhk-agent</li>
    <li>uhk-udev-rules</li>
  </ul>
</details>

File: nixos/modules/hardware/keyboard/uhk.nix [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+    You may want to install the uhk-agent package.
+    '';
+
+  };
+
+  config = mkIf cfg.enable {
+    services.udev.packages = [ pkgs.uhk-udev-rules ];
+  };
+}
+
```
Comment (SuperSandro2000) [inline nixos/modules/hardware/keyboard/uhk.nix:22]: That is one new line to much.

File: nixos/modules/hardware/keyboard/uhk.nix [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+let
+  cfg = config.hardware.keyboard.uhk;
+in
+{
+  options.hardware.keyboard.uhk = {
+    enable = mkEnableOption ''
+    non-root access to the firmware of UHK keyboards.
+    You need it when you want to flash a new firmware on the keyboard.
+    Access to the keyboard is granted to users in the "input" group.
+    You may want to install the uhk-agent package.
```
Comment (SuperSandro2000) [inline nixos/modules/hardware/keyboard/uhk.nix:13]:
```suggestion
      non-root access to the firmware of UHK keyboards.
      You need it when you want to flash a new firmware on the keyboard.
      Access to the keyboard is granted to users in the "input" group.
      You may want to install the uhk-agent package.
```

File: pkgs/os-specific/linux/uhk-udev-rules/default.nix [hunk 8]
```diff
@@ -0,0 +1,22 @@
+{ lib, stdenv, uhk-agent }:
+
+stdenv.mkDerivation {
+  pname = "uhk-udev-rules";
+  version = lib.getVersion uhk-agent;
```
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/uhk-udev-rules/default.nix:5]:
```suggestion
  version = uhk-agent.version;
```

File: pkgs/os-specific/linux/uhk-udev-rules/default.nix [hunk 9]
```diff
@@ -0,0 +1,22 @@
...
+    runHook preInstall
+    install -D -m 644 ${uhk-agent.out}/rules/50-uhk60.rules $out/lib/udev/rules.d/50-uhk60.rules
+    runHook postInstall
+  '';
+
+  meta = {
+    description = ''
+    udev rules for UHK keyboards
+    from https://ultimatehackingkeyboard.com.
+    '';
```
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/uhk-udev-rules/default.nix:18]:
```suggestion
    description = "udev rules for UHK keyboards from https://ultimatehackingkeyboard.com";
```

File: pkgs/os-specific/linux/uhk-agent/default.nix [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+
+    install -m 444 -D ${appimageContents}/${pname}.desktop -t $out/share/applications
+    install -m 644 -D ${appimageContents}/resources/rules/50-uhk60.rules $out/rules/50-uhk60.rules
+    substituteInPlace $out/share/applications/${pname}.desktop \
+      --replace 'Exec=AppRun' 'Exec=${pname}'
+    cp -r ${appimageContents}/usr/share/icons $out/share
+  '';
+
+  meta = with lib; {
+    description = "Agent is the configuration application of the Ultimate Hacking Keyboard.";
```
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/uhk-agent/default.nix:34]:
```suggestion
    description = "Agent is the configuration application of the Ultimate Hacking Keyboard";
```

File: pkgs/os-specific/linux/uhk-agent/default.nix [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+    name = "${name}.AppImage";
+    sha256 = "sha256-t2Jwd/x0eTZ4xBaCb/FomH/zSRLt7IIERUF9n9ONCpE=";
+  };
+
+  appimageContents = appimageTools.extract {
+    name = "${pname}-${version}";
+    inherit src;
+  };
+in appimageTools.wrapType2 {
+  inherit src name;
```
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/uhk-agent/default.nix:19]:
```suggestion
  inherit pname version src;
```

File: pkgs/os-specific/linux/uhk-agent/default.nix [hunk 12]
```diff
@@ -0,0 +1,40 @@
+{ appimageTools, lib, fetchurl, polkit, udev }:
+
+# Credits: https://cloud.tissot.de/gitea/benneti/nixos/src/branch/main/pkgs/uhk-agent/default.nix
+let
+  pname = "uhk-agent";
+  version = "1.5.15";
+  name = "${pname}-${version}";
```
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/uhk-agent/default.nix:7]:
```suggestion
```

File: pkgs/os-specific/linux/uhk-agent/default.nix [hunk 13]
```diff
@@ -0,0 +1,40 @@
+{ appimageTools, lib, fetchurl, polkit, udev }:
+
+# Credits: https://cloud.tissot.de/gitea/benneti/nixos/src/branch/main/pkgs/uhk-agent/default.nix
+let
```
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/uhk-agent/default.nix:4]:
```suggestion

let
```

**Comment by poelzi** [comment]
damnit. now i also packaged it because this is not merged and I did not see this PR :(

File: pkgs/os-specific/linux/uhk-agent/default.nix [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+    name = "${name}.AppImage";
+    sha256 = "sha256-nJxfhmHt2bnu+lQOL7zQPzBe03n6yAVtyG7Y1atOpgg=";
+  };
+
+  appimageContents = appimageTools.extract {
+    name = "${pname}-${version}";
+    inherit src;
+  };
+in appimageTools.wrapType2 {
+  inherit pname src name;
```
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/uhk-agent/default.nix:17]:
```suggestion
  inherit name pname version src;
```

I think we need this to make ``uhk-agent.version`` work.

Comment (ngiger) [reply pkgs/os-specific/linux/uhk-agent/default.nix:17]: I added the version to the meta attributes of uhk-agent. Now I can set `version = uhk-agent.meta.version;` in uhk-udev-rules/default.nix.

Hope this this okay for your.

File: pkgs/os-specific/linux/uhk-agent/default.nix [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+
+  extraPkgs = pkgs: with pkgs; [ polkit udev ];
+
+  extraInstallCommands = ''
+    mv $out/bin/${name} $out/bin/${pname}
```
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/uhk-agent/default.nix:22]:
```suggestion
    mv $out/bin/${pname}-${version} $out/bin/${pname}
```

File: pkgs/os-specific/linux/uhk-agent/default.nix [hunk 16]
```diff
@@ -0,0 +1,39 @@
...
+    substituteInPlace $out/share/applications/${pname}.desktop \
+      --replace 'Exec=AppRun' 'Exec=${pname}'
+    cp -r ${appimageContents}/usr/share/icons $out/share
+  '';
+
+  meta = with lib; {
+    description = "Agent is the configuration application of the Ultimate Hacking Keyboard";
+    homepage = "https://github.com/UltimateHackingKeyboard/agent";
+    license = licenses.unfreeRedistributable;
+    version = version;
```
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/uhk-agent/default.nix:35]:
```suggestion
```

File: pkgs/os-specific/linux/uhk-udev-rules/default.nix [hunk 17]
```diff
@@ -0,0 +1,19 @@
+{ lib, stdenv, uhk-agent }:
+
+stdenv.mkDerivation {
+  pname = "uhk-udev-rules";
+  version = uhk-agent.meta.version;
```
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/uhk-udev-rules/default.nix:5]:
```suggestion
  inherit (uhk-agent)version;

```

File: pkgs/os-specific/linux/uhk-agent/default.nix [hunk 18]
```diff
@@ -0,0 +1,37 @@
+{ appimageTools, lib, fetchurl, polkit, udev }:
+let
+  pname = "uhk-agent";
+  version = "1.5.17";
+  src = fetchurl {
+    url = "https://github.com/UltimateHackingKeyboard/agent/releases/download/v${version}/UHK.Agent-${version}-linux-x86_64.AppImage";
+    name = "${name}.AppImage";
```
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/uhk-agent/default.nix:7]:
```suggestion
    name = "${pname}-${version}.AppImage";
```

File: pkgs/os-specific/linux/uhk-udev-rules/default.nix [hunk 19]
```diff
@@ -0,0 +1,20 @@
+{ lib, stdenv, uhk-agent }:
+
+stdenv.mkDerivation {
+  pname = "uhk-udev-rules";
+  inherit (uhk-agent)version;
```
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/uhk-udev-rules/default.nix:5]:
```suggestion
  inherit (uhk-agent) version;
```

File: pkgs/os-specific/linux/uhk-agent/default.nix [hunk 20]
```diff
@@ -0,0 +1,37 @@
...
+  extraInstallCommands = ''
+    mv $out/bin/${pname}-${version} $out/bin/${pname}
+
+    install -m 444 -D ${appimageContents}/${pname}.desktop -t $out/share/applications
+    install -m 644 -D ${appimageContents}/resources/rules/50-uhk60.rules $out/rules/50-uhk60.rules
+    substituteInPlace $out/share/applications/${pname}.desktop \
+      --replace 'Exec=AppRun' 'Exec=${pname}'
+    cp -r ${appimageContents}/usr/share/icons $out/share
+  '';
+
```
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/uhk-agent/default.nix:29]:
```suggestion
  '';
  
  # wrapType2 does not passthru pname+version
  passthru.version = version;

```

File: pkgs/os-specific/linux/uhk-agent/default.nix [hunk 21]
```diff
@@ -0,0 +1,40 @@
...
```
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/uhk-agent/default.nix:29]:
```suggestion

```

**Comment by SuperSandro2000** [comment]
@ofborg eval

**Comment by ngiger** [comment]
Thanks for your commits. As a newcomer I did not know, that I first have to pull in your changes and then to try to fix things.
Anyway, I am happy that things are starting to look fine.

**Comment by SuperSandro2000** [comment]
It doesn't really matter if you apply my changes or do them locally. I just tried to speed things up so we can get this merged.

**Pull request merged**